### PR TITLE
[Spaces] Context switching

### PR DESCRIPTION
**Pull request by SimonBrandner**
Fixes [#16730](https://github.com/vector-im/element-web/issues/16730)

Peek 2021-03-26 13-55

### CODE REVIEW

File: src/stores/SpaceStore.tsx [hunk 1]
```diff
@@ -35,12 +35,17 @@ import {objectDiff} from "../utils/objects";
...
 
 type SpaceKey = string | symbol;
 
 interface IState {}
 
 const ACTIVE_SPACE_LS_KEY = "mx_active_space";
+export const LAST_VIEWED_ROOMS = "mx_last_viewed_rooms";
+
+// We can't use HOME_SPACE here because JSON.stringify() will ignore any Symbols
+export const LAST_VIEWED_ROOMS_HOME = "home_space";
```
Comment (SimonBrandner) [inline src/stores/SpaceStore.tsx:48]: There might be a better solution that I don't know of here

File: src/stores/SpaceStore.tsx [hunk 2]
```diff
@@ -111,6 +116,18 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
...
+        const spaceId = space?.roomId || LAST_VIEWED_ROOMS_HOME;
+        const lastViewedRooms = JSON.parse(window.localStorage.getItem(LAST_VIEWED_ROOMS));
+        const roomId = lastViewedRooms[spaceId];
+
+        if (roomId) {
+            dis.dispatch({
+                action: "view_room",
+                room_id: roomId,
+            });
+        } // TODO: Handle else
```
Comment (SimonBrandner) [inline src/stores/SpaceStore.tsx:129]: Is there any desired behaviour for this?

Comment (t3chguy) [reply src/stores/SpaceStore.tsx:129]: in the case of no prior room I think it should go to space home, or for the home space just to the first room in the room list maybe

File: src/stores/SpaceStore.tsx [hunk 3]
```diff
@@ -41,6 +41,9 @@ type SpaceKey = string | symbol;
 interface IState {}
 
 const ACTIVE_SPACE_LS_KEY = "mx_active_space";
+const LAST_VIEWED_ROOMS = "mx_last_viewed_rooms";
+
+const LAST_VIEWED_ROOMS_HOME = "home_space";
 
 export const HOME_SPACE = Symbol("home-space");
```
Comment (SimonBrandner) [inline src/stores/SpaceStore.tsx:48]: Should I try to use the `Symbol`?

Comment (t3chguy) [reply src/stores/SpaceStore.tsx:48]: if you're putting the room id in the localStorage keys you could do 

`mx_......_${s === HOME_SPACE ? "home" : s.roomId)` or similar

Comment (SimonBrandner) [reply src/stores/SpaceStore.tsx:48]: > `mx_......_${s === HOME_SPACE ? "home" : s.roomId} or similar`

I am already doing something like this:

```ts
const activeSpaceId = this.activeSpace?.roomId || LAST_VIEWED_ROOMS_HOME;                
window.localStorage.setItem(`${LAST_VIEWED_ROOMS}_${activeSpaceId}`, payload.room_id);
```

Am I missing something?

File: src/stores/SpaceStore.tsx [hunk 4]
```diff
@@ -41,6 +41,9 @@ type SpaceKey = string | symbol;
 interface IState {}
 
 const ACTIVE_SPACE_LS_KEY = "mx_active_space";
+const LAST_VIEWED_ROOMS = "mx_last_viewed_rooms";
```
Comment (t3chguy) [inline src/stores/SpaceStore.tsx:44]: can we turn this into a small method to get the LS key based on the given space as we have the same concat happening in two places e.g `getLastViewedRoomsStorageKey(space?)` or something more brief

File: src/stores/SpaceStore.tsx [hunk 5]
```diff
@@ -111,6 +114,26 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
         this.emit(UPDATE_SELECTED_SPACE, this.activeSpace);
         this.emit(SUGGESTED_ROOMS, this._suggestedRooms = []);
 
+        // view last selected room from space
+        const spaceId = space?.roomId || LAST_VIEWED_ROOMS_HOME;
+        const roomId = window.localStorage.getItem(`${LAST_VIEWED_ROOMS}_${spaceId}`);
+
+        if (roomId) {
+            defaultDispatcher.dispatch({
```
Comment (t3chguy) [inline src/stores/SpaceStore.tsx:122]: we should probably check that we are still joined to the room, if we are not then maybe go to space home

File: src/stores/SpaceStore.tsx [hunk 6]
```diff
@@ -467,6 +490,15 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
             case "view_room": {
                 const room = this.matrixClient?.getRoom(payload.room_id);
 
+                // persist last viewed room from a space
+
+                // We don't want to save if the room is a
+                // space room since it can cause problems
```
Comment (t3chguy) [inline src/stores/SpaceStore.tsx:496]: whats the problem is can cause?

Comment (SimonBrandner) [reply src/stores/SpaceStore.tsx:496]: This is weird - I remember there being a problem but I can't repro it now. If any issues arise we know where to look

File: src/stores/SpaceStore.tsx [hunk 7]
```diff
@@ -51,6 +54,11 @@ export const UPDATE_SELECTED_SPACE = Symbol("selected-space");
 
 const MAX_SUGGESTED_ROOMS = 20;
 
+const getLastViewedRoomsStorageKey = (spaceId?) => {
```
Comment (t3chguy) [inline src/stores/SpaceStore.tsx:57]: I meant more like it takes a `space?: Room` and then does `space?.roomId || LAST_VIEWED_ROOMS_HOME;` internally, sorry about not being clear

Comment (SimonBrandner) [reply src/stores/SpaceStore.tsx:57]: Yeah, I just realized that. No worries, I don't seem to be functioning very well today... :sweat_smile:

File: src/stores/SpaceStore.tsx [hunk 8]
```diff
@@ -111,6 +117,25 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
         this.emit(UPDATE_SELECTED_SPACE, this.activeSpace);
         this.emit(SUGGESTED_ROOMS, this._suggestedRooms = []);
 
+        // view last selected room from space
+        const roomId = window.localStorage.getItem(getLastViewedRoomsStorageKey(this.activeSpace));
+
+        if (roomId && this.matrixClient.getRoom(roomId).getMyMembership() === "join") {
```
Comment (t3chguy) [inline src/stores/SpaceStore.tsx:123]:
```suggestion
        if (roomId && this.matrixClient?.getRoom(roomId)?.getMyMembership() === "join") {
```

**Review by t3chguy — APPROVED** [review]
Great, thanks for this !

**Comment by tanriol** [comment]
After the latest update I'm seeing `develop.element.io` automatically switching from the space I'm trying to enter. Can this room switch cause a space switch to happen after it?

UPD: now my `develop.element.io` page is stuck switching back and forth between two spaces and rooms without any user input :-)
@SimonBrandner @t3chguy

**Pull request merged**